src/pause.sol

```diff
@@ -74,30 +74,18 @@ contract DSPause {
 
     function execute(address guy, bytes memory data, uint when)
         public
-        returns (bytes memory response)
+        returns (bytes memory)
     {
         bytes32 id = tag(guy, data, when);
 
         require(now >= when + delay, "ds-pause: delay not passed");
         require(scheduled[id] == true, "ds-pause: unscheduled execution");
 
         scheduled[id] = false;
+        (bool success, bytes memory response) = guy.delegatecall.gas(gasleft() - 1250)(data);
```

Comment:
The tests fail with `OutOfGas` errors if I don't have this `.gas(gasleft() - 1250)` statement.

I don't understand why? How is gas forwarded here? Does `delegatecall` consume all the gas that it is passed? I couldn't see anything in the docs describing how this works.... :disappointed:

Reply:
I guess it has something to do with [this](https://github.com/ethereum/wiki/wiki/Subtleties):

> A CALL or CREATE can consume at most 63/64 of the gas remaining at the time the CALL is made; if a CALL asks for more than this prescribed maximum, then the inner call will only have the prescribed maximum gas regardless of how much gas was asked for.

`gasleft() / 64 * 63` works in fact.

It's not clear to me though why this doesn't just silently succeed, since it should provide the called contract the prescribed maximum anyway.


So what would be happening is that if you don't specify (at least) `gasleft() - 1250`, the called contract gets

Reply:
As far as I understand the gas you rest from the `gasleft()` is the necessary to execute the `delegatecall` function itself (add, sub, mload operations). Otherwise you would be forwarding all the gas to execute the delegated code and nothing remains for doing the previous operations. In `ds-proxy` we are leaving 5K gas for this to have an important gap just in case in the future the gas costs increment.
There was some discussion about it in this PR: https://github.com/dapphub/ds-proxy/pull/18 )(I was actually confused about its purpose in that moment).

Reply:
[This](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md) is the actual EIP:

> If a call asks for more gas than the maximum allowed amount (i.e. the total amount of gas remaining in the parent after subtracting the gas cost of the call and memory expansion), do not return an OOG error; instead, if a call asks for more gas than all but one 64th of the maximum allowed amount, call with all but one 64th of the maximum allowed amount of gas (this is equivalent to a version of EIP-901 plus EIP-1142). CREATE only provides all but one 64th of the parent gas to the child call.

So it seems to me that hevm should not return an OOG error here.

I think this whole 63/64 thing is meant to address security concerns, and replaces the stack limit:

> replacing the call stack depth limit with a "softer" gas-based restriction, thereby eliminating call stack depth attacks as a class of attack that contract developers have to worry about and hence increasing contract programming safety.